### PR TITLE
Feature/#81 Add Google Auth Strategy 🚀

### PR DESCRIPTION
**Pull request by ArturoGrullon**
Conveniently found this [guide](https://dev.to/imichaelowolabi/how-to-implement-login-with-google-in-nest-js-2aoa), adapted it to our project, and it's working like a charm. 

Didn't know if we have a dedicated account or not so I set up one in the mid-time.

Just pulling from google at the moment: 

- Name 
- Last name 
- Email 
- Photo 
- Secret key

### CODE REVIEW

**Comment by Kerosz** [comment]
@ArturoDucasse I've looked into the article you linked, and it's a good starting point but that is not going to work with our already established `JWT` system. I'm gonna try to give you some suggestions while I review the code.

File: .env.example [hunk 1]
```diff
@@ -12,4 +12,9 @@ AUTH_PEPPER=secretPepper
 
 # DB Options
 DB_URI=yourMongoUri
-TEST_DB_URI=yourTestMongoUri
\ No newline at end of file
+TEST_DB_URI=yourTestMongoUri
+
+#Google auth
+# Todo: Change to production credentials
+GOOGLE_CLIENT_ID= 114619963004-2mec4e5gicsjgn83ufq0b9cudd0apgkp.apps.googleusercontent.com
```
Comment (Kerosz) [inline .env.example]: Make sure you integrate these new env variables into our config system

File: src/auth/auth.controller.ts [hunk 2]
```diff
@@ -37,4 +39,15 @@ export class AuthController {
   public async signUpWithEmailAndPassword(@Body() createUserDto: CreateUserDto) {
     return this.authService.signUpWithEmailAndPassword(createUserDto);
   }
+
+  @Get("google")
+  @UseGuards(AuthGuard("google"))
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  async googleAuth(@Req() _req: any) {}
```
Comment (Kerosz) [inline src/auth/auth.controller.ts]: Remove the eslint comment and just add  a void return, and type `req` to use express `Request` object

`import { Request } from "express";`

```suggestion
  async googleAuth(@Req() _req: Request) {
      return;
  }
```

File: src/auth/auth.controller.ts [hunk 3]
```diff
@@ -37,4 +39,15 @@ export class AuthController {
...
+
+  @Get("google/redirect")
+  @UseGuards(AuthGuard("google"))
+  googleAuthRedirect(@Req() req: any) {
+    return this.authService.googleLogin(req);
```
Comment (Kerosz) [inline src/auth/auth.controller.ts]: We could rename this to something a bit more descriptive

`import { Profile } from "passport-google-oauth20";`

```suggestion
    return this.authService.loginWithGoogleProvider(req.user as Profile);
```

File: src/auth/auth.service.ts [hunk 4]
```diff
@@ -40,6 +33,22 @@ export class AuthService {
     return userRecordWithoutPassword;
   }
 
+  private signUser(id: string, role: RoleType) {
```
Comment (Kerosz) [inline src/auth/auth.service.ts]: While you are here, you could give this function a better name, and update all references

```suggestion
  private signUserToken(id: string, role: RoleType) {
```

File: src/auth/auth.service.ts [hunk 5]
```diff
@@ -40,6 +33,22 @@ export class AuthService {
...
+    return this.jwtService.sign({
+      sub: id,
+      claim: role,
+    });
+  }
+
+  public googleLogin(req: any) {
```
Comment (Kerosz) [inline src/auth/auth.service.ts]: Use same name as in the controller `loginWithGoogleProvider`

```suggestion
  public loginWithGoogleProvider(providerUserProfile: Profile) {
```

File: src/auth/auth.service.ts [hunk 6]
```diff
@@ -40,6 +33,22 @@ export class AuthService {
...
+    if (!req.user) return "No user from google";
+
+    //TODO?: Save user in database
```
Comment (Kerosz) [inline src/auth/auth.service.ts]: Yep we need to save the user into the DB. To do this we need to make some changes to the `UserEntity` model, to add probably 2 more fields, those being `provider` and `providerId`. Then we can save the user and return the `userRecord` from DB along with `JWT` token

PS: I have coded this function on my local when I pulled down your changes so if you want we can have a talk later on VC.

File: src/auth/strategies/google.strategy.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import { PassportStrategy } from "@nestjs/passport";
+import { Strategy, VerifyCallback } from "passport-google-oauth20";
+import { config } from "dotenv";
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: We should use our own `configService`

File: src/auth/strategies/google.strategy.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+import { Injectable } from "@nestjs/common";
+
+config();
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: remove dotenv config function

File: src/auth/strategies/google.strategy.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+@Injectable()
+export class GoogleStrategy extends PassportStrategy(Strategy, "google") {
+  constructor() {
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: Inject our own `configService` after you update it to include the 2 env variables you added

```suggestion
  constructor(readonly configService: ConfigService) {
```

File: src/auth/strategies/google.strategy.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    super({
+      clientID: process.env.GOOGLE_CLIENT_ID,
+      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+      callbackURL: "http://localhost:3000/api/v1/auth/google/redirect",
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: use injected config for the variables

```suggestion
      clientID: configService.authOptions.googleClientId,
      clientSecret: configService.authOptions.googleClientSecret,
      callbackURL: `${configService.apiOptions.hostname}/${configService.apiOptions.prefix}/v1/auth/google/redirect`,
```

File: src/auth/strategies/google.strategy.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+      scope: ["email", "profile"],
+    });
+  }
+
+  async validate(
+    accessToken: string,
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: We are not going to use google accessToken, as we are using our `JWT` implementation.

```suggestion
    _accessToken: string,
```

File: src/auth/strategies/google.strategy.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    _refreshToken: string,
+    profile: any,
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: You can type this as `Profile`

`import {  Profile } from "passport-google-oauth20";`

```suggestion
    profile: Profile,
```

File: src/auth/strategies/google.strategy.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    done: VerifyCallback
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: We don't need to verify any CB in here, just to pass the google profile to the controller

```suggestion

```

File: src/auth/strategies/google.strategy.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+  ): Promise<any> {
+    const { name, emails, photos } = profile;
+    const user = {
+      email: emails[0].value,
+      firstName: name.givenName,
+      lastName: name.familyName,
+      picture: photos[0].value,
+      accessToken,
+    };
+    done(null, user);
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: we just need to return the profile

```suggestion
    return profile;
```

File: src/auth/strategies/google.strategy.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+      scope: ["email", "profile"],
+    });
+  }
+
+  async validate(
+    accessToken: string,
+    _refreshToken: string,
+    profile: any,
+    done: VerifyCallback
+  ): Promise<any> {
```
Comment (Kerosz) [inline src/auth/strategies/google.strategy.ts]: type the `any` to use the `Profile` type

```suggestion
  ): Promise<Profile> {
```

File: src/user/models/user.entity.ts [hunk 16]
```diff
@@ -45,12 +46,15 @@ export class UserEntity extends BaseEntitySchema {
   public role!: RoleType;
 
   @Prop({ type: String, default: null })
+  @Exclude()
```
Comment (Kerosz) [inline src/user/models/user.entity.ts]: Remove all the attempts to serialize the entity including imports from this PR as we have a separate PR for that issue.

File: src/user/user.controller.ts [hunk 17]
```diff
@@ -49,6 +51,7 @@ export class UserController {
   @ApiNotFoundResponse({ description: "The record was not found." })
   @ApiForbiddenResponse({ description: "Forbidden!" })
   @UseRoles(Role.ADMIN)
+  @UseInterceptors(ClassSerializerInterceptor)
```
Comment (Kerosz) [inline src/user/user.controller.ts]: Same as in the `UserEntity` remove all attempts to serialize the responses, including imports as we have a separate issue for it

File: src/auth/auth.controller.ts [hunk 18]
```diff
@@ -37,4 +39,15 @@ export class AuthController {
...
   }
+
+  @Get("google")
+  @UseGuards(AuthGuard("google"))
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  async googleAuth(@Req() _req: any) {}
+
+  @Get("google/redirect")
+  @UseGuards(AuthGuard("google"))
+  googleAuthRedirect(@Req() req: any) {
```
Comment (Kerosz) [inline src/auth/auth.controller.ts]: type `req` to use the express `Request` object

`import { Request } from "express";`

```suggestion
  googleAuthRedirect(@Req() req: Request) {
```

**Review by Kerosz — CHANGES_REQUESTED** [review]
Great work, coming with all this code so far, you are getting close, and nicely done for setting up the google account for it :)

Some things to highlight from the review I left down below, I have a local implementation of the function in auth service `loginWithGoogleProvider`, so if you would like to cross reference it, we can have a talk on VC.

I have seen you somehow generated a `package-lock.json`. Make sure to delete it and only use `yarn` to install deps, so we only have `yarn.lock`.

File: .env.example (hunk 1 above)
Comment (ArturoGrullon) [reply .env.example]: I did? my bad, forgot to delete the lock file after the installation (was lazy to search for the global install in yarn 😄). 

I will add those to the config system, and sure, later on we can cross-reference it

File: src/auth/auth.controller.ts (hunk 2 above)
Comment (ArturoGrullon) [reply src/auth/auth.controller.ts]: Will do

File: src/auth/auth.controller.ts (hunk 3 above)
Comment (ArturoGrullon) [reply src/auth/auth.controller.ts]: Yeah, it does seem more cleaner that way. 
Thanks for the suggestion

File: src/auth/auth.service.ts (hunk 4 above)
Comment (ArturoGrullon) [reply src/auth/auth.service.ts]: Will do

File: src/auth/auth.service.ts (hunk 5 above)
Comment (ArturoGrullon) [reply src/auth/auth.service.ts]: Will do

File: src/auth/auth.service.ts (hunk 6 above)
Comment (ArturoGrullon) [reply src/auth/auth.service.ts]: Got it, I saw the previous message. Later on, we could chat if you are available

File: src/auth/strategies/google.strategy.ts (hunk 7 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Will do, saw it but was confused so just went for the typical approach 😄

File: src/auth/strategies/google.strategy.ts (hunk 8 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Will do

File: src/auth/strategies/google.strategy.ts (hunk 9 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Will do

File: src/auth/strategies/google.strategy.ts (hunk 10 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Will do

File: src/auth/strategies/google.strategy.ts (hunk 11 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Got it,

File: src/auth/strategies/google.strategy.ts (hunk 12 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Yeah, will add this

File: src/auth/strategies/google.strategy.ts (hunk 13 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Got it

File: src/auth/strategies/google.strategy.ts (hunk 14 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Hehe, got it. 

Much cleaner this way

File: src/auth/strategies/google.strategy.ts (hunk 15 above)
Comment (ArturoGrullon) [reply src/auth/strategies/google.strategy.ts]: Will make this change

File: src/user/models/user.entity.ts (hunk 16 above)
Comment (ArturoGrullon) [reply src/user/models/user.entity.ts]: Oh, I totally unnoticed this commit come along. 

Thanks for pointing it out

File: src/user/user.controller.ts (hunk 17 above)
Comment (ArturoGrullon) [reply src/user/user.controller.ts]: Will do

File: src/auth/auth.controller.ts (hunk 18 above)
Comment (ArturoGrullon) [reply src/auth/auth.controller.ts]: Will do

**Comment by ArturoGrullon** [comment]
Hey, thanks for the comments kero. 

I appreciate the time you took to read it through, will make these changes. Yet I haven't done much, just followed the guide and connecting the dots (barely 😄) but I'm feeling more comfortable in Nest... at least I'm not that completely lost 😂.

I will get to these suggestions right away.

**Comment by Kerosz** [comment]
@all-contributors please add @ArturoDucasse  for code

**Review by Kerosz — APPROVED** [review]

**Comment by Kerosz** [comment]
@ArturoDucasse seems like I cannot request you for a review as it's your PR. I have pushed the changes we talked about and the ones I left in my review. You can look over the files to see if there is something else I've missed

**Pull request merged**